README.md
Sprout
===========
[![Build Status](https://travis-ci.org/levigroker/Sprout.svg)](https://travis-ci.org/levigroker/Sprout)
[![Version](http://img.shields.io/cocoapods/v/Sprout.svg)](http://cocoapods.org/?q=Sprout)
[![Platform](http://img.shields.io/cocoapods/p/Sprout.svg)]()
[![License](http://img.shields.io/cocoapods/l/Sprout.svg)](https://github.com/levigroker/Sprout/blob/master/LICENSE.txt)

Use to bootstrap the (excellent) [CocoaLumberjack](https://github.com/robbiehanson/CocoaLumberjack) logging framework and add additional functionality such as Custom Loggers, log archiving, crash stack traces, and more.

### Installing

If you're using [CocoPods](http://cocopods.org) it's as simple as adding this to your `Podfile`:

	pod 'Sprout', '~> 3.1'

Sprout makes use of some preprocessor defines to configure the logging level and some functionality. These preprocessor definitions need to be added to the Pods target for Sprout, as opposed to your own project build settings, because the Pods library gets compiled without being exposed to your project build settings. To do this, you can add a `post_install` hook to your `Podfile` (as seen below).

* `DEBUG=1` If defined, this sets the default logging level to be verbose (`ddLogLevel = DDLogLevelVerbose`) and enables the TTY (console) logger,
otherwise the default is the warning level (`ddLogLevel = DDLogLevelWarning`) and no TTY logger.
* `SPROUT_LOG_LEVEL` can be used to override the default log level. Define `SPROUT_LOG_LEVEL` to whatever log level is appropriate for your configuration. See the **Log Levels** section below.
* `SPROUT_DISABLE_DYNAMIC_LOG_LEVEL=1` By default, Sprout supports CocoaLumberjack's dynamic log level usage by declaring `ddLogLevel` as `const` (`static const int ddLogLevel`). If you don't need dynamic log level support, and would like the extra speed disabling it will provide, you can disable this by defining `SPROUT_DISABLE_DYNAMIC_LOG_LEVEL=1`

#### Podfile post_install

Here's an example `post_install` hook which adds the `DEBUG` and `SPROUT_DISABLE_DYNAMIC_LOG_LEVEL` preprocessor definitions to the `Pods-Sprout` target of the `Pods` project. *Note* `DEBUG` is not added to the `Release` configuration.
This also overrides the default log level by setting `SPROUT_LOG_LEVEL` to a different log level for `Release` vs. other Schemes. 

		post_install do |installer_representation|

		  # Grab the `project` object from the installer (cocoapods < 0.38 use `project`, cocoapods >= 0.38 use pods_project)
		  # See https://github.com/CocoaPods/CocoaPods/issues/2292
		  if installer_representation.respond_to?(:project)
			project = installer_representation.project
		  else
			project = installer_representation.pods_project
		  end

		  # Set our default log levels
		  sprout = (project.targets.select { |target| target.name == 'Sprout' }).first
		  if sprout
			sprout.build_configurations.each do |config|
			  config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= ['$(inherited)']
			  if config.name == 'Release'
				config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] << 'SPROUT_LOG_LEVEL=DDLogLevelWarning'
			  else
				config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] << 'SPROUT_LOG_LEVEL=DDLogLevelVerbose'
				if !config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'].include? 'DEBUG=1'
				  config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] << 'DEBUG=1'
				end
			  end
			end
		  end

		end
	
### Documentation

 In the simplest case, setup is just:

* Add `<Sprout/Sprout.h>` to your precompiled header (or import it directly wherever you use it):


		#ifdef __OBJC__
		#import <Foundation/Foundation.h>
		//Third Party
		#import <Sprout/Sprout.h>
		#endif

* For iOS, start Logging in your `application:didFinishLaunchingWithOptions:` UIApplicationDelegate

		- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
		{
			//Initialize logging
			[[Sprout sharedInstance] startLogging];
			...

* For MacOS, start Logging in your `applicationWillFinishLaunching:` NSApplicationDelegate

		- (void)applicationWillFinishLaunching:(NSNotification *)notification
		{
			//Initialize logging
			[[Sprout sharedInstance] startLogging];
			...


* Define `DEBUG`
 In your Preprocessor Macros target build settings, define `DEBUG=1` (this may not be needed, as it is a default setting in later Xcode project templates).

* _Run!_

After the above setup, you should be able to run and see:

		CocoaLumberjack loggers initialized!

appear in your console.

#### Log Levels

By default, Sprout allows the use of dynamic log levels, meaning `setLogLevel:` can be sent at runtime to set the desired log level. This comes with a slight performance hit for log entries, and can be disabled by defining `SPROUT_DISABLE_DYNAMIC_LOG_LEVEL=1`. If disabled, the log level will be static and can only be set at compile time.

The log level defaults to `DDLogLevelVerbose` if `DEBUG` is defined and set to a non-zero value. If `DEBUG` is not defined (or set to zero) the log level defaults to `DDLogLevelWarning`.

The default log level can be overridden by defining `SPROUT_LOG_LEVEL` and setting it to the desired log level.

NOTE: If you're using Sprout with CocoaPods, simply defining this in your precompiled header or project build settings will not have the desired affect, since Sprout is compiled into the Pods library before these are traversed by the pre-compiler. So you will need to define `SPROUT_LOG_LEVEL` in the Podfile `post_install` hook.

See the **Podfile post_install** section above for an example `post_install` hook which does this.

#### Default Loggers

Sprout has default loggers which will be installed under certain circumstances.

* A file logger (`DDFileLogger`) will always be installed. This logger has 24 hour rolling and maximum seven log files.
* A TTY logger (`DDTTYLogger`) will be installed if `DEBUG=1` is true.

You can override which loggers get installed by supplying a `loggersBlock` and returning the loggers you desire before calling `startLogging`. i.e. you can add additional loggers, or remove (some of) the default loggers passed to the block.

Additionally, you can use `addLogger:`, `addLogger:withLogLevel:`, `removeLogger:`, and `removeAllLoggers` after the call to `startLoggers` to modify which loggers are installed.

#### Custom Log Formatter

Sprout comes with `SproutCustomLogFormatter` which outputs two lines for every log entry. For example:

		2014-05-20 16:06:45:602         <60b> startLogging(Sprout.m 164)
		2014-05-20 16:06:45:602  [INfO] CocoaLumberjack loggers initialized!

* The first line specifies the *thread* (`<60b>`), the *function* (`startLogging`), the *file* (`Sprout.m`) and *line number* within the file (`164`).
* The second line is the *actual log message* (`CocoaLumberjack loggers initialized!`), prefixed by the *log level* (`[INfO]`) of the message.

Both lines are prefixed by a *date/time stamp* (`2014-05-20 16:06:45:602`).

If you wish to supply your own log formatter you can provide a `logFormatterBlock` which will be used to obtain a `DDLogFormatter` to use for each logger. The block must be set before calling `startLogging`.

#### Crashlytics Usage

[Crashlytics](http://crashlytics.com) logging is supported by Sprout.

Sadly, using Cocoapods to automatically manage this dependency is not presently possible when `use_frameworks!` is specified. This is due to Crashlytics being provided as a static library (see [this thread](https://twittercommunity.com/t/crashlytics-cocoapod-dynamic-framework-support/47000)).

However, you can manually add support for logging to Crashlytics by making use of the 
 `CrashlyticsLogger` implementation provided. You'll need to include `CrashlyticsLogger.h/m` directly in your project and add it to Sprout before calling `startLogging`

__NOTE:__ You should initialize Crashlytics (call `[Fabric with:@[Crashlytics.class]]`) inside Sprout `startLogging:` block handler. 

Example:

    [Sprout sharedInstance].loggersBlock = ^NSSet<id<DDLogger>> *(NSSet<id<DDLogger>> *defaultLoggers) {
    	NSMutableSet *retVal = [NSMutableSet setWithSet:defaultLoggers];
    	[retVal addObject:[CrashlyticsLogger sharedInstance]];
    	return retVal;
    };

    [[Sprout sharedInstance] startLogging:^{
	    [Fabric with:@[Crashlytics.class]]
    	[[Sprout sharedInstance] logAppAndDeviceInfo];
    }];

#### App Center Usage

[App Center](https://appcenter.ms) logging is supported by Sprout.

You can manually add support for logging to App Center by making use of the 
 `AppCenterLogger` implementation provided. You'll need to include `AppCenterLogger.h/m` directly in your project and add it to Sprout before calling `startLogging`

Example:

    [Sprout sharedInstance].loggersBlock = ^NSSet<id<DDLogger>> *(NSSet<id<DDLogger>> *defaultLoggers) {
    	NSMutableSet *retVal = [NSMutableSet setWithSet:defaultLoggers];
    	[retVal addObject:[AppCenterLogger sharedInstance]];
    	return retVal;
    };

    [[Sprout sharedInstance] startLogging:^{
    	[[Sprout sharedInstance] logAppAndDeviceInfo];
    }];

### Licence

This work is licensed under the [Creative Commons Attribution 4.0 International License](https://creativecommons.org/licenses/by/4.0/).
Please see the included [LICENSE.txt](https://github.com/levigroker/Sprout/blob/master/LICENSE.txt) for complete details.

### About
A professional iOS engineer by day, my name is Levi Brown. Authoring a blog
[grokin.gs](http://grokin.gs), I am reachable via:

Twitter [@levigroker](https://twitter.com/levigroker)  
Email [levigroker@gmail.com](mailto:levigroker@gmail.com)  

Your constructive comments and feedback are always welcome.


CHANGELOG.md
* 1.0 - March 1, 2013
 * Initial public release.
* 1.1 - September 20, 2013
  * Added dynamic log level support.
  * Added basic app, device, and OS logging functionality.
* 1.2 - February 13, 2014
  * Fixing use of TestFlightLogger
* 1.2.1 - March 11, 2014
  * Fixing deadlock issue when logging exceptions.
* 1.3 - May 20, 2014
  * Making Sprout "cross platform" (iOS and OSX).
  * Added documentation for Podfile `post_install` hook and preprocessor definitions.
  * Added documentation for `CustomLogFormatter` format.
* 2.0 - August 12, 2014
  * Added Crashlytics logging support.
  * Now dynamically install optional framework loggers (TestFlight, Crashlytics)
  * Removed now unused `TESTFLIGHT` preprocessor directive.
  * Added capability to override the default loggers.
  * Added capability to override the default log formatter.
  * Renamed `CustomLogFormatter` to `SproutCustomLogFormatter`.
  * Added documentation for new features.
* 2.0.1 - August 13, 2014
  * Adding provisions for calling any previously installed exception handler.
  * Avoiding namespace conflicts for exception and signal handlers.
* 2.0.2 - August 13, 2014
  * Adding empty strings to TestFlight weak constants to avoid weak linking issues.
* 2.0.3 - October 3, 2014
  * Adding ability to override default static log level by defining `SPROUT_LOG_LEVEL`.
  * Cleaning up preprocessor macro checking and defines.
  * Adding Sprout-internal log category.
* 2.0.4 - October 17, 2014
  * Armoring against NULL backtrace symbols.
* 2.0.5 - February 5, 2015
  * Fixing TARGET_OS_MAC conditional compilation.
  * Adding `deviceName` to `logAppAndDeviceInfo` output.
* 2.0.6 - February 19, 2015
  * Adding `deviceMachine` to `logAppAndDeviceInfo` output.
* 2.0.7 - March 17, 2015
  * Removing `TestFlightLogger` as it is no longer relevant.
  * Exposed additional internals such as app and device info methods.
* 2.0.8 - June 26, 2015
  * Reimplementing backtrace functions.
  * Adding tests for backtrace funciton.
* 2.0.9 - October 16, 2015
  * Changing build project and directory structure.
  * Updated copyright and Readme.
  * Added Travis.ci build/test automation.
* 2.0.10 - November 9, 2015
  * Adding `SproutDDLogAdditions.h` with synchronous and asynchronous "log always" macros.
* 2.1 - December 7, 2015
  * Removing `logsAsZippedData` and dependency on [SSZipArchive](https://github.com/ZipArchive/ZipArchive).
* 2.2 - August 4, 2016
  * Adding code to condense instances of the same kind of signal into one, within a 1 second threshold.
* 3.0 - December 28, 2016
  * Updated to use CocoaLumberjack 3.0
  * Updated license.
* 3.1 - February 22, 2017
  * Provided block based configuration (instead of overrides)
  * Updated `logFiles` method to return NSURLs (**breaking API change**)
  * Removed weak dependency on Crashlytics for logging.
  * CocoaLumberjack 3.1 dependency
* 3.1.1 - March 20, 2019
  * Added `AppCenterLogger.h/m` for logging to [App Center](https://appcenter.ms)\
* 3.1.2 - July 9, 2021
  * Incorporated PR from @gyratorycircus to "log always" app/device info at startup.
* 3.2 - 2021-09-21
  * Updated AppCenter API use to 4.3. Using DDOSLogger instead of DDTTYLogger.
* 3.2.1 - 2021-09-21
  * Updating macOS deployment target to 10.15.
* 3.2.2 - 2021-10-11
  * Adding support for WatchOS
* 3.2.3 - 2021-10-17 - @gyratorycircus
  * Skips install of fatal signal handlers which are not supported on watchOS

Sprout/Sprout.m
//
//  Sprout.m
//
//  Created by Levi Brown on October 4, 2012.
//  Copyright (c) 2012-2017 Levi Brown <mailto:levigroker@gmail.com> This work is
//  licensed under the Creative Commons Attribution 4.0 International License. To
//  view a copy of this license, visit https://creativecommons.org/licenses/by/4.0/
//  or send a letter to Creative Commons, PO Box 1866, Mountain View, CA 94042, USA.
//
//  The above attribution and the included license must accompany any version of
//  the source code, binary distributable, or derivatives.
//

#include <execinfo.h>
#import <sys/sysctl.h>

#if TARGET_OS_IPHONE
#if TARGET_OS_WATCH
#import <WatchKit/WatchKit.h>
#else // iOS
#import <UIKit/UIKit.h>
#endif
#else //Mac
#import <SystemConfiguration/SCDynamicStoreCopySpecific.h>

static NSString * const kSystemVersionPlistLocation = @"/System/Library/CoreServices/SystemVersion.plist";
static NSString * const kSystemVersionKeyProductName = @"ProductName";
static NSString * const kSystemVersionKeyProductUserVisibleVersion = @"ProductUserVisibleVersion";
static NSString * const kSystemVersionKeyProductBuildVersion = @"ProductBuildVersion";

#endif //TARGET

static NSString * const kSysInfoKeyHardwarePlatform = @"hw.model";
static NSString * const kSysInfoKeyHardwareMachine = @"hw.machine";
static NSTimeInterval const kSignalReportingThresholdTimeInterval = 1.0f;

#import "Sprout.h"
#import "SproutCustomLogFormatter.h"

#define DDLogException(frmt, ...)   LOG_MAYBE(NO, ddLogLevel, DDLogFlagError, 0, nil, "Exception Handler", frmt, ##__VA_ARGS__)
#define DDLogSignal(frmt, ...)      LOG_MAYBE(NO, ddLogLevel, DDLogFlagError, 0, nil, "Signal Handler", frmt, ##__VA_ARGS__)

#define SproutLogError(frmt, ...)   LOG_MAYBE(NO,                sproutInternalLogLevel, DDLogFlagError,   SPROUT_LOG_CONTEXT, nil, __PRETTY_FUNCTION__, @"[Sprout] "frmt, ##__VA_ARGS__)
#define SproutLogWarn(frmt, ...)    LOG_MAYBE(LOG_ASYNC_ENABLED, sproutInternalLogLevel, DDLogFlagWarning, SPROUT_LOG_CONTEXT, nil, __PRETTY_FUNCTION__, @"[Sprout] "frmt, ##__VA_ARGS__)
#define SproutLogInfo(frmt, ...)    LOG_MAYBE(LOG_ASYNC_ENABLED, sproutInternalLogLevel, DDLogFlagInfo,    SPROUT_LOG_CONTEXT, nil, __PRETTY_FUNCTION__, @"[Sprout] "frmt, ##__VA_ARGS__)
#define SproutLogDebug(frmt, ...)   LOG_MAYBE(LOG_ASYNC_ENABLED, sproutInternalLogLevel, DDLogFlagDebug,   SPROUT_LOG_CONTEXT, nil, __PRETTY_FUNCTION__, @"[Sprout] "frmt, ##__VA_ARGS__)
#define SproutLogVerbose(frmt, ...) LOG_MAYBE(LOG_ASYNC_ENABLED, sproutInternalLogLevel, DDLogFlagVerbose, SPROUT_LOG_CONTEXT, nil, __PRETTY_FUNCTION__, @"[Sprout] "frmt, ##__VA_ARGS__)

void sproutExceptionHandler(NSException *exception);
void sproutSignalHandler(int signal);
NSUncaughtExceptionHandler *priorHandler;
NSTimeInterval thresholdSignalTime = DBL_MAX;
NSString *lastSignal = nil;

@interface Sprout ()

@property (nonatomic,assign) BOOL started;
@property (nonatomic,strong) DDFileLogger *fileLogger;
@property (nonatomic,strong) NSMutableOrderedSet *startupMessageBlocks;

@end

//Original exception and signal handling concept and code from http://www.cocoawithlove.com/2010/05/handling-unhandled-exceptions-and.html
//Since modified

void sproutExceptionHandler(NSException *exception)
{
    DDLogException(@"Exception: %@\nCall stack:\n%@", exception, [exception callStackSymbols]);
    if (priorHandler && priorHandler != &sproutExceptionHandler)
    {
        (*priorHandler)(exception);
    }
}

void sproutSignalHandler(int signal)
{
    NSString *signalSring = @"unknown";
    switch (signal)
    {
        case SIGABRT:
            signalSring = @"SIGABRT";
            break;
        case SIGILL:
            signalSring = @"SIGILL";
            break;
        case SIGBUS:
            signalSring = @"SIGBUS";
            break;
        case SIGSEGV:
            signalSring = @"SIGSEGV";
            break;
        default:
            break;
    }
    
    //Collapse similar signals within a given time threshold
    if ([lastSignal isEqualToString:signalSring] && thresholdSignalTime > [NSDate timeIntervalSinceReferenceDate])
    {
        //The same kind of signal was reported within the threshold time, so we ignore it.
    }
    else
    {
        //The signal does not match our last signal kind (or there was no last signal), so report it
        //OR
        //The threshold time has passed, so report the signal again
        
        //Update the threshold time
        thresholdSignalTime = [NSDate timeIntervalSinceReferenceDate] + kSignalReportingThresholdTimeInterval;

        NSString *backtrace = [Sprout backtraceSkipping:2 /* Skip the backrace method and this signal handler */ length:20 /* Get the next 20 stack frames */];
        DDLogSignal(@"Recieved %@ signal. Call stack:\n%@", signalSring, backtrace);
    }
    
    //Track the current signal kind
    lastSignal = signalSring;
}

@implementation Sprout

#pragma mark - Lifecycle

+ (Sprout *)sharedInstance
{
    static dispatch_once_t onceQueue;
    static Sprout *sprout = nil;
    
    dispatch_once(&onceQueue, ^{ sprout = [[self alloc] init]; });
    return sprout;
}

- (id)init
{
    if ((self = [super init]))
    {
		_startupMessageBlocks = [[NSMutableOrderedSet alloc] init];
    }
    
    return self;
}

#pragma mark Class Level

+ (NSString *)backtraceSkipping:(NSUInteger)skip length:(NSUInteger)length
{
    NSArray *trimmedBacktrace = [self trimmedBacktraceSkipping:skip length:length];
    NSString *retVal = [trimmedBacktrace componentsJoinedByString:@"\n"];
    return retVal;
}

+ (NSArray *)trimmedBacktraceSkipping:(NSUInteger)skip length:(NSUInteger)length
{
    NSMutableArray *retVal = [NSMutableArray array];
    if (length > 0)
    {
        NSArray *backtrace = [NSThread callStackSymbols];
        NSUInteger finalIndex = (length - 1) + skip;
        [backtrace enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
            if (idx >= skip)
            {
                [retVal addObject:obj];
            }
            *stop = idx >= finalIndex;
        }];
    }
    
    return retVal;
}

+ (NSString *)stringForLogLevel:(int)logLevel
{
    NSString *retVal = nil;
    
    switch (logLevel) {
        case DDLogLevelOff:
            retVal = @"DDLogLevelOff";
            break;
        case DDLogLevelError:
            retVal = @"DDLogLevelError";
            break;
        case DDLogLevelWarning:
            retVal = @"DDLogLevelWarning";
            break;
        case DDLogLevelInfo:
            retVal = @"DDLogLevelInfo";
            break;
        case DDLogLevelDebug:
            retVal = @"DDLogLevelDebug";
            break;
        case DDLogLevelVerbose:
            retVal = @"DDLogLevelVerbose";
            break;
        case DDLogLevelAll:
            retVal = @"DDLogLevelAll";
            break;
        default:
            retVal = @"<Unkown>";
            break;
    }
    
    return retVal;
}

#pragma mark - Implementation

- (void)startLogging
{
	[self startLogging:nil];
}

- (void)startLogging:(void(^)(void))completion
{
    if (!self.started)
    {
        //Initialize logging
        
        //Store any prior handler so we can chain call to it.
        priorHandler = NSGetUncaughtExceptionHandler();
        //Register our handler for uncaught exceptions.
        NSSetUncaughtExceptionHandler(&sproutExceptionHandler);
        
        //Register our handler for signals
		//Note: Skipped on watchOS where "sigaction on fatal signals is not supported"
#if !TARGET_OS_WATCH
        //Create the signal action structure
        struct sigaction signalAction;
        //Initialize the signal action structure
        memset(&signalAction, 0, sizeof(signalAction));
        //Set 'signalHandler' as the handler in the signal action structure
        signalAction.sa_handler = &sproutSignalHandler;
        //Set 'signalHandler' as the handlers for SIGABRT, SIGILL, SIGBUS, and SIGSEGV
        sigaction(SIGABRT, &signalAction, NULL);
        sigaction(SIGILL, &signalAction, NULL);
        sigaction(SIGBUS, &signalAction, NULL);
        sigaction(SIGSEGV, &signalAction, NULL);
#endif
		
        [self addDefaultLoggers];

        BOOL defaultLogLevel = YES;
        #ifdef SPROUT_LOG_LEVEL
        defaultLogLevel = NO;
        #endif
        
        BOOL dynamicLogLevel = YES;
        #if SPROUT_DISABLE_DYNAMIC_LOG_LEVEL
        dynamicLogLevel = NO;
        #endif

		__weak typeof(self) weakSelf = self;
		[self addStartupMessageBlock:^{
			SproutLogDebug(@"Log level %@ '%@'. Dynamic log level is %@abled.", defaultLogLevel ? @"defaulted to" : @"set by 'SPROUT_LOG_LEVEL' to", [weakSelf.class stringForLogLevel:LOG_LEVEL_DEF], dynamicLogLevel ? @"en" : @"dis");
		}];
		
		[self addStartupMessageBlock:^{
			SproutLogInfo(@"CocoaLumberjack loggers initialized!");
		}];
		
		if (completion)
		{
			completion();
		}
		
		[self logStartupMessageBlocks];

		self.started = YES;
    }
}

#if !SPROUT_DISABLE_DYNAMIC_LOG_LEVEL
- (void)setLogLevel:(int)logLevel
{
    ddLogLevel = logLevel;
}
#endif

- (void)logAppAndDeviceInfo
{
    NSString *appName = [self appName];
    NSString *appVersion = [self appVersion];
    NSString *appIdentifier = [self appIdentifier];
    NSString *appBuildNumber = [self appBuildNumber];
    NSString *deviceName;
    NSString *deviceModel;
    NSString *deviceMachine;
    NSString *systemName;
    NSString *systemVersion;
    
    [self deviceName:&deviceName deviceModel:&deviceModel deviceMachine:&deviceMachine systemName:&systemName systemVersion:&systemVersion];

    DDLogAlwaysInfo(@"%@ %@ (%@ %@) %@ \"%@\" (%@), %@ %@", appName ?: @"<unknown_app_name>", appVersion ?: @"<unknown_app_version>", appIdentifier ?: @"<unknown_app_identifier>", appBuildNumber ?: @"<unknown_app_build_number>", deviceModel ?: @"<unknown_device_model>", deviceName ?: @"<unknown_device_name>", deviceMachine ?: @"<unknown_device_machine>", systemName ?: @"<unknown_system_name>", systemVersion ?: @"<unknown_system_version>");
}

- (NSArray *)logFiles
{
    NSUInteger maximumLogFilesToReturn = MIN(self.fileLogger.logFileManager.maximumNumberOfLogFiles, 10);
    NSMutableArray *logFiles = [NSMutableArray arrayWithCapacity:maximumLogFilesToReturn];
    NSArray *sortedLogFileInfos = [self.fileLogger.logFileManager sortedLogFileInfos];
    for (int i = 0; i < MIN(sortedLogFileInfos.count, maximumLogFilesToReturn); ++i)
    {
        DDLogFileInfo *logFileInfo = [sortedLogFileInfos objectAtIndex:i];
		NSURL *logFileURL = logFileInfo.filePath ? [NSURL fileURLWithPath:logFileInfo.filePath isDirectory:NO] : nil;
        if (logFileURL)
        {
            [logFiles addObject:logFileURL];
        }
    }
    return logFiles;
}

#pragma mark - Logging Setup

- (void)addDefaultLoggers
{
	NSSet<id<DDLogger>> *loggers = [self defaultLoggers];
	
	if (self.loggersBlock)
	{
		loggers = self.loggersBlock(loggers);
	}

	for (id<DDLogger> logger in loggers)
    {
        [self addLogger:logger];
		[self addStartupMessageBlock:^{
			SproutLogVerbose(@"Added logger: '%@'", [logger loggerName]);
		}];
    }
}

- (NSSet<id<DDLogger>> *)defaultLoggers
{
	NSMutableSet *loggers = [NSMutableSet setWithCapacity:2];
	
	id<DDLogger> logger = nil;
	
	logger = [self setupOSLogger];
	if (logger)
	{
		[loggers addObject:logger];
	}
	
	logger = [self setupFileLogger];
	if (logger)
	{
		self.fileLogger = logger;
		[loggers addObject:logger];
	}
	
    return loggers;
}

- (void)addLogger:(id <DDLogger>)logger
{
    [self addLogger:logger withLogLevel:DDLogLevelAll];
}

- (void)addLogger:(id <DDLogger>)logger withLogLevel:(NSUInteger)logLevel
{
    if (logger)
    {
		id<DDLogFormatter> formatter = [[SproutCustomLogFormatter alloc] init];
		
		if (self.logFormatterBlock)
		{
			formatter = self.logFormatterBlock(formatter);
		}

		[logger setLogFormatter:formatter];
        
        [DDLog addLogger:logger];
    }
}

- (void)removeLogger:(id <DDLogger>)logger
{
    [DDLog removeLogger:logger];
}

- (void)removeAllLoggers
{
    [DDLog removeAllLoggers];
}

- (NSArray *)allLoggers
{
    return [DDLog allLoggers];
}

- (DDOSLogger *)setupOSLogger
{
	DDOSLogger *logger = nil;
    
    #if SPROUT_CONSOLE_LOGGING
    //Console
    logger = [DDOSLogger sharedInstance];
    #endif
    
    return logger;
}

- (DDFileLogger *)setupFileLogger
{
    DDFileLogger *logger = nil;
    
    #if SPROUT_FILE_LOGGING
    //File logging
    logger = [[DDFileLogger alloc] init];
    logger.rollingFrequency = 60 * 60 * 24; // 24 hour rolling
    logger.logFileManager.maximumNumberOfLogFiles = 7;
    #endif

    return logger;
}

#pragma mark - Helpers

- (void)addStartupMessageBlock:(void (^__nonnull)(void))messageBlock
{
	if (messageBlock)
	{
		[self.startupMessageBlocks addObject:[messageBlock copy]];
	}
}

- (void)logStartupMessageBlocks
{
	for (void (^messageBlock)(void) in self.startupMessageBlocks)
	{
		messageBlock();
	}
}

- (NSString *)appVersion
{
    NSString *retVal = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleShortVersionString"];
    return retVal;
}

- (NSString *)appBuildNumber
{
    NSString *retVal = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleVersion"];
    return retVal;
}

- (NSString *)appName
{
    NSDictionary *info = [[NSBundle mainBundle] infoDictionary];
    NSString *name = [info objectForKey:@"CFBundleDisplayName"];
    if (!name)
    {
        name = [info objectForKey:@"CFBundleName"];
    }

    return name;
}

- (NSString *)appIdentifier
{
    NSString *retVal = [[NSBundle mainBundle] bundleIdentifier];
    return retVal;
}

#pragma mark - Device Info

- (void)deviceName:(NSString **)deviceName deviceModel:(NSString **)deviceModel deviceMachine:(NSString **)deviceMachine systemName:(NSString **)systemName systemVersion:(NSString **)systemVersion
{
#if TARGET_OS_IPHONE
#if TARGET_OS_WATCH
	[self watchDeviceName:deviceName deviceModel:deviceModel deviceMachine:deviceMachine systemName:systemName systemVersion:systemVersion];
#else // iOS
    [self iosDeviceName:deviceName deviceModel:deviceModel deviceMachine:deviceMachine systemName:systemName systemVersion:systemVersion];
#endif
#elif TARGET_OS_MAC
    [self osxDeviceName:deviceName deviceModel:deviceModel deviceMachine:deviceMachine systemName:systemName systemVersion:systemVersion];
#endif
}

#if TARGET_OS_IPHONE

#if TARGET_OS_WATCH
- (void)watchDeviceName:(NSString **)deviceName deviceModel:(NSString **)deviceModel deviceMachine:(NSString **)deviceMachine systemName:(NSString **)systemName systemVersion:(NSString **)systemVersion
{
	if (deviceName)
	{
		*deviceName = [WKInterfaceDevice currentDevice].name;
	}

	if (deviceModel)
	{
		*deviceModel = [WKInterfaceDevice currentDevice].model;
	}

	if (deviceMachine)
	{
		*deviceMachine = [self sysInfoByName:kSysInfoKeyHardwareMachine];
	}

	if (systemName)
	{
		*systemName = [WKInterfaceDevice currentDevice].systemName;
	}

	if (systemVersion)
	{
		*systemVersion = [WKInterfaceDevice currentDevice].systemVersion;
	}
}
#else // iOS
- (void)iosDeviceName:(NSString **)deviceName deviceModel:(NSString **)deviceModel deviceMachine:(NSString **)deviceMachine systemName:(NSString **)systemName systemVersion:(NSString **)systemVersion
{
    if (deviceName)
    {
        *deviceName = [UIDevice currentDevice].name;
    }
    
    if (deviceModel)
    {
        *deviceModel = [UIDevice currentDevice].model;
    }
    
    if (deviceMachine)
    {
        *deviceMachine = [self sysInfoByName:kSysInfoKeyHardwareMachine];
    }
    
    if (systemName)
    {
        *systemName = [UIDevice currentDevice].systemName;
    }
    
    if (systemVersion)
    {
        *systemVersion = [UIDevice currentDevice].systemVersion;
    }
}
#endif

#else // Mac OS

- (void)osxDeviceName:(NSString **)deviceName deviceModel:(NSString **)deviceModel deviceMachine:(NSString **)deviceMachine systemName:(NSString **)systemName systemVersion:(NSString **)systemVersion
{
    if (deviceName)
    {
        *deviceName = CFBridgingRelease(SCDynamicStoreCopyComputerName(NULL, NULL));
    }
    
    if (deviceModel)
    {
        *deviceModel = [self sysInfoByName:kSysInfoKeyHardwarePlatform];
    }
    
    if (deviceMachine)
    {
        *deviceMachine = [self sysInfoByName:kSysInfoKeyHardwareMachine];
    }
    
    //See http://stackoverflow.com/questions/11072804/mac-os-x-10-8-replacement-for-gestalt-for-testing-os-version-at-runtime
    NSDictionary *systemVersionDict = [NSDictionary dictionaryWithContentsOfFile:kSystemVersionPlistLocation];
    
    if (systemName)
    {
        *systemName = systemVersionDict[kSystemVersionKeyProductName];
    }
    
    if (systemVersion)
    {
        NSString *version = systemVersionDict[kSystemVersionKeyProductName];
        NSString *buildNumber = systemVersionDict[kSystemVersionKeyProductBuildVersion];
        *systemVersion = [NSString stringWithFormat:@"%@ (%@)", version ?: @"<unknown_system_version", buildNumber ?: @"<unknown_system_build_number>"];
    }
}

#endif //TARGET

- (NSString *)sysInfoByName:(NSString *)name
{
    NSString *retVal = nil;
    
    if (name.length > 0)
    {
        const char *cName = [name cStringUsingEncoding:NSUTF8StringEncoding];
        
        size_t size;
        sysctlbyname(cName, NULL, &size, NULL, 0);
        char *buffer = malloc(size);
        sysctlbyname(cName, buffer, &size, NULL, 0);
        retVal = [NSString stringWithCString:buffer encoding:NSUTF8StringEncoding];
        free(buffer);
    }
    
    return retVal;
}

- (NSURL *)tempDirectory
{
    //Modified from: http://cocoawithlove.com/2009/07/temporary-files-and-folders-in-cocoa.html
    
    NSURL *retVal = nil;
    
    NSString *tempDirectoryTemplate = [NSTemporaryDirectory() stringByAppendingPathComponent:[[NSUUID UUID] UUIDString]];
    const char *tempDirectoryTemplateCString = [tempDirectoryTemplate fileSystemRepresentation];
    char *tempDirectoryNameCString = (char *)malloc(strlen(tempDirectoryTemplateCString) + 1);
    strcpy(tempDirectoryNameCString, tempDirectoryTemplateCString);
    
    char *result = mkdtemp(tempDirectoryNameCString);
    if (result)
    {
        retVal = [NSURL fileURLWithFileSystemRepresentation:tempDirectoryTemplateCString isDirectory:YES relativeToURL:nil];
    }
    
    free(tempDirectoryNameCString);
    
    return retVal;
}

@end
